### ISSUE TITLE
Check for interfaces instead of concrete types

### ISSUE THREAD

**Issue by misja** [akesterson/dpath-python#35]
I was hoping dpath would work with objects implementing e.g. `collections.MutableMapping` but unfortunately it will only work with concrete types like `dict`, `list` and `tuple`.

```python
import dpath
from collections import Mapping, MutableMapping

class Namespace(MutableMapping):
    def __init__(self, data={}):
        self._mapping = {}
        self._mapping.update(data)

    def __len__(self):
        return len(self._mapping)

    def __iter__(self):
        return iter(self._mapping)

    def __contains__(self, key):
        return key in self._mapping

    def __getitem__(self, key):
        return self._mapping[key]

    def __setitem__(self, key, value):
        self._mapping[key] = value

    def __delitem__(self, key):
        del self._mapping[key]

x = Namespace()
x['user'] = Namespace()
x['user']['group'] = 'test'

assert isinstance(dict(), Mapping) == True
assert isinstance({}, Mapping) == True

assert x['user']['group'] == 'test'
assert dpath.util.get(x, '/user/group') == 'test'
```

**Comment by akesterson** [comment]
This particular test case wouldn't work, because dpath also needs to be able to path into lists and tuples. And for those, this assertion fails:

    >>> isinstance([], Mapping)       
    False
    >>> isinstance(tuple([]), Mapping)
    False

... But I agree we can do a better job of caring about the types of objects we operate on. If it can index like a duck, insert like a duck and iterate like a duck, we should treat it like a duck. I'll think more on this and get back to you.

@calebcase thoughts?

**Comment by misja** [comment]
Of course it will not work, since lists and tuples are of type `collections.Sequence`

```python
from collections import Sequence

isinstance([], Sequence)
isinstance(tuple([]), Sequence)
```

**Comment by misja** [comment]
Er, didn't mean to be blunt there, just that I think that all checks could be reduced to types `collection.Mapping` and `collections.Sequence` :)

**Comment by akesterson** [comment]
Fair point, I wasn't aware of collections.Sequence (or most of collections.* , for that matter).

I'm pretty busy this week so I likely won't get around to writing up a branch to test this for several days, if you're okay waiting. If not, there are only 33 instances of issubclass/isinstance that I can find, and I'd welcome a pull request that solves this problem and passes unit tests.

**Comment by calebcase** [comment]
Sounds good to me!

**Comment by akesterson** [comment]
Closed by merge, released in 1.4.0.
